test/bdd/pkg/agent/agent_controller_steps.go

```diff
@@ -0,0 +1,313 @@
+		bddContext:     ctx,
+		Args:           make(map[string]string),
+		ControllerURLs: make(map[string]string),
+		WebhookURLs:    make(map[string]string),
+		webSocketConns: make(map[string]*websocket.Conn),
+	}
+}
+
+// RegisterSteps registers agent steps.
+func (a *Steps) RegisterSteps(s *godog.Suite) {
+	s.Step(`^"([^"]*)" agent is running on "([^"]*)" port "([^"]*)" with controller "([^"]*)"$`,
+		a.validateAgentConnection)
+	s.Step(`^"([^"]*)" responds to connect request from Issuer adapter \("([^"]*)"\) within "([^"]*)" seconds$`,
+		a.handleDIDConnectRequest)
+}
+
+func (a *Steps) validateAgentConnection(agentID, inboundHost,
+	inboundPort, controllerURL string) error {
+	if err := a.checkAgentIsRunning(agentID, controllerURL); err != nil {
+		return err
+	}
+
+	// verify inbound
+	if err := a.healthCheck(fmt.Sprintf("http://%s:%s", inboundHost, inboundPort)); err != nil {
+		logger.Debugf("Unable to reach inbound '%s' for agent '%s', cause : %s", controllerURL, agentID, err)
+		return err
+	}
+
+	logger.Debugf("Agent '%s' running inbound on '%s' and port '%s'", agentID, inboundHost, inboundPort)
+
+	return nil
+}
+
+func (a *Steps) checkAgentIsRunning(agentID, controllerURL string) error {
+	// verify controller
+	err := a.healthCheck(controllerURL)
+	if err != nil {
+		logger.Debugf("Unable to reach controller '%s' for agent '%s', cause : %s", controllerURL, agentID, err)
+		return err
+	}
+
+	logger.Debugf("Agent '%s' running controller '%s'", agentID, controllerURL)
+
+	a.ControllerURLs[agentID] = controllerURL
+
+	// create and register websocket connection for notifications
+	u, err := url.Parse(controllerURL)
+	if err != nil {
+		return fmt.Errorf("invalid controller URL [%s]", controllerURL)
+	}
+
+	wsURL := fmt.Sprintf("ws://%s%s/ws", u.Host, u.Path)
+
+	ctx, cancel := goctx.WithTimeout(goctx.Background(), timeoutWS)
+	defer cancel()
+
+	conn, _, err := websocket.Dial(ctx, wsURL, nil) //nolint:bodyclose
+	if err != nil {
+		return fmt.Errorf("failed to dial connection from '%s' : %w", wsURL, err)
+	}
+
+	a.webSocketConns[agentID] = conn
+
+	return nil
+}
+
+func (a *Steps) healthCheck(endpoint string) error {
+	if strings.HasPrefix(endpoint, "http") {
+		resp, err := http.Get(endpoint) //nolint: gosec
+		if err != nil {
+			return err
+		}
+
+		err = resp.Body.Close()
+		if err != nil {
+			logger.Errorf("Failed to close response body : %s", err)
+		}
+
+		return nil
+	}
+
+	return errors.New("url scheme is not supported for url = " + endpoint)
+}
+
+func (a *Steps) handleDIDConnectRequest(agentID, issuerID string, timeout int) error {
+	// Mock CHAPI request from Issuer
+	invitationJSON := a.bddContext.Store[bddutil.GetDIDConectRequestKey(issuerID, agentID)]
+
+	connectionID, err := a.receiveInvitation(agentID, invitationJSON)
+	if err != nil {
+		return err
+	}
+
+	err = a.approveInvitation(agentID)
+	if err != nil {
+		return err
+	}
+
+	// Added to mock CHAPI timeout (ie, DIDExchange should happen with this duration)
```

Comment:
not really "timeout" - more like wallet is doing stuff (interacting with the issuer and creating VCs) before responding